File: 2018/JavaScript/Day01/program.js
#!/usr/bin/env osascript -l JavaScript

var app = Application.currentApplication();
app.includeStandardAdditions = true;

function* infinite(xs) {
  const len = xs.length;
	for (let i = 0; true; i = (i + 1) % len) {
    yield xs[i];
  }
}

function calcResult1(changes) {
  return changes.reduce((a,b) => a + b, 0);
}

function calcResult2(frequencyChanges) {
  var infiniteChanges = infinite(frequencyChanges);
  var lastFrequency = 0;
  var previousFrequencies = new Set([lastFrequency]);
  for (var changeInFrequency of infiniteChanges) {
    var nextFrequency = lastFrequency + changeInFrequency;
    if (previousFrequencies.has(nextFrequency)) {
      return nextFrequency;
    } else {
      previousFrequencies.add(nextFrequency);
      lastFrequency = nextFrequency;
    }
  }
}

function run(argv) {
  console.log("Running ...");

  var input = app.read(Path('./input.txt'))
                .split('\n')
                .map(x => x.trim())
                .filter(x => x != '')
                .map(x => parseInt(x));

  console.log('Result 1:', calcResult1(input));
  console.log('Result 2:', calcResult2(input));
}

File: 2018/Ruby/Day01/program.rb
require_relative 'frequency_summer'
require_relative 'duplicate_frequency_finder'

class App
  def load_input(path)
    IO.readlines(path).map(&:strip).reject(&:empty?).map(&:to_i)
  end

  def run
    input = load_input('./input.txt')
    
    result1 = FrequencySummer.new.calc(input)
    puts "Result 1: #{result1}"

    result2 = DuplicateFrequencyFinder.new.calc(input)
    puts "Result 2: #{result2}"
  end
end

app = App.new
app.run
